### PR TITLE
WIP: support for reading VTK_POLYGON

### PR DESCRIPTION
**Pull request by eolianoe**
Following #322, here is a first draft to read VTK files with polygons inside. The current code does not support cells with a custom number of points and the "easy" way to support it, is to loop over the cells the fill the `numnodes` and `offsets` arrays.

The `cell_data` variable is not filled yet as I'm looking for a nice solution.

### CODE REVIEW

**Comment by nschloe** [comment]
Looks interesting!

It'd be a good idea to add writing for polygons as well, so that this part of the code can be tested like the rest.

Small nitpick: Logical variables (like `polygon` in your draft) should have names indicating the Booleanness, e.g., `is_polygon` or `has_polygon`.

**Comment by eolianoe** [comment]
The polygons can now be written, but cell_data is not done as the dictionnay of dictionnaries is a bit strange for me now.

**Comment by nschloe** [comment]
I'm backpacking right now, but will look at it once I'm back.

**Comment by eolianoe** [comment]
You closed #322, but I didn't get any feedback after your trip. Can I have your opinion on this PR?

**Comment by nschloe** [comment]
Sorry, missed this one! Looking at it now.

**Comment by nschloe** [comment]
Okay, so I think this is a valuable addition and your coding skill + familiarity with meshio is good enough for us to get somewhere.

What I don't like about the approach is that there's basically a separate `if` branch whenever polygons are involved. Perhaps there's no way around this because polygons can have any number of nodes, so this may be okay.

What we definitely need here are tests. Tests in meshio work in such a way that they first _write_ some data to a file, _read_ it afterwards, and the output data is then compared with the one we started with. So the next thing we need is a support for polygon writing here. Do you think you can do it? If yes, I could add some tests. After that, we could fix the code style here and there (nothing major) and merge. What do you think?

**Comment by eolianoe** [comment]
The only way I see to remove the `if` is to use a list of list rather than an array of list to store the connectivity but it may break the others readers/writers.

If I'm not mistaken, the writer is working fine with the current code. For the tests, is it better to write one polygon or a set of polygons and triangles/quads ?

No problem for the code style :)

File: meshio/helpers.py [hunk 1]
```diff
@@ -205,7 +205,8 @@ def write(filename, mesh, file_format=None, **kwargs):
 
     # check cells for sanity
     for key, value in mesh.cells.items():
-        assert value.shape[1] == num_nodes_per_cell[key]
+        if "polygon" not in key:
```
Comment (nschloe) [inline meshio/helpers.py]: This should be `key != "polygon"` to make sure the rest is tested.

File: meshio/vtk_io.py [hunk 2]
```diff
@@ -359,25 +359,59 @@ def translate_cells(data, types, cell_data_raw):
...
+    has_polygon = meshio_to_vtk_type["polygon"] in bins
 
     # Deduct offsets from the cell types. This is much faster than manually
     # going through the data array. Slight disadvantage: This doesn't work for
     # cells with a custom number of points.
     numnodes = numpy.empty(len(types), dtype=int)
-    for tpe, idx in bins.items():
-        numnodes[idx] = vtk_type_to_numnodes[tpe]
-    offsets = numpy.cumsum(numnodes + 1) - (numnodes + 1)
+    if not has_polygon:
```
Comment (nschloe) [inline meshio/vtk_io.py]: Negated `if`s with an `else`, e.g.,
```
if not hungry:
    don't eat
else:
    eat
```
are harder to read than
```
if hungry:
    eat
else:
    don't eat
```
Please adapt all conditionals in the PR accordingly.

File: meshio/vtk_io.py [hunk 3]
```diff
@@ -359,25 +359,59 @@ def translate_cells(data, types, cell_data_raw):
...
+        for tpe, idx in bins.items():
+            numnodes[idx] = vtk_type_to_numnodes[tpe]
+        offsets = numpy.cumsum(numnodes + 1) - (numnodes + 1)
+    else:
+        # If some polygons are in the VTK file, loop over the cells
+        nbcells = len(types)
+        offsets = numpy.empty(len(types), dtype=int)
+        offsets[0] = 0
+        for idx in numpy.arange(nbcells - 1):
```
Comment (nschloe) [inline meshio/vtk_io.py]: `numpy.arange` -> `range` (everywhere)

Comment (nschloe) [reply meshio/vtk_io.py]: We should also think about the possibility of removing this loop. It's gonna be super slow.

File: meshio/vtk_io.py [hunk 4]
```diff
@@ -472,16 +506,27 @@ def _write_cells(f, cells, write_binary):
...
         for key in cells:
-            d = numpy.full(len(cells[key]), meshio_to_vtk_type[key]).astype(
-                numpy.dtype(">i4")
-            )
+            if "polygon" not in key:
+                d = numpy.full(len(cells[key]), meshio_to_vtk_type[key]).astype(
+                    numpy.dtype(">i4")
+                )
+            else:
+                d = numpy.full(len(cells[key]), meshio_to_vtk_type["polygon"]).astype(
```
Comment (nschloe) [inline meshio/vtk_io.py]: This looks like it's not going to work if both `polygon` and other cells are in the mesh.

File: meshio/helpers.py (hunk 1 above)
Comment (eolianoe) [reply meshio/helpers.py]: Yes, but I do not want the rest to be tested as, for now, the key is "polygonN" with N the number of vertices/edges

File: meshio/vtk_io.py (hunk 4 above)
Comment (eolianoe) [reply meshio/vtk_io.py]: why? It seems to work with my first tests

File: meshio/helpers.py (hunk 1 above)
Comment (nschloe) [reply meshio/helpers.py]: If the key `polygonN`, then this conditional is always true. I think I misunderstand something here.

File: meshio/vtk_io.py (hunk 4 above)
Comment (nschloe) [reply meshio/vtk_io.py]: If the `"polygon"` key _is_ in `cells` (along with other keys), then you always pick `meshio_to_vtk_type["polygon"]`. That's the part that looks suspicious to me.

Comment (eolianoe) [reply meshio/vtk_io.py]: Nope, there is loop over the keys and the test is made inside the loop

File: meshio/helpers.py (hunk 1 above)
Comment (eolianoe) [reply meshio/helpers.py]: Yes, that's the point all polygon have the same VTK_TYPE

Comment (nschloe) [reply meshio/helpers.py]: Ah, okay. For clarity, this should be replaced by `key[:7] == "polygon"`.

**Comment by eolianoe** [comment]
Here is a file to test the current code, it writes a pentagonal prism with some quadtrilaterals and triangles faces, then read it.
```
#!/usr/bin/env python3

import meshio
import numpy as np
import math

DIM = 3


def readMesh(filename):
    mesh = meshio.read(filename)
    points = mesh.points
    elements = mesh.cells
    return points, elements


def prism(n):
    h = 1.0
    radius = 0.5/np.sin(np.pi/n)

    theta = np.linspace(0.0, 2.0*np.pi, n+1)
    verts = np.zeros(shape=(2*n, DIM))
    for i in np.arange(0, n):
        verts[i, 0] = radius*math.cos(theta[i])
        verts[i, 1] = radius*math.sin(theta[i])
        verts[i, 2] = 0.
        verts[n+i, 0] = radius*math.cos(theta[i])
        verts[n+i, 1] = radius*math.sin(theta[i])
        verts[n+i, 2] = h

    faces = []
    polygonalFaces = np.zeros(shape=(2, n), dtype=int)
    for i in np.arange(0, n):
        faces.append([i, (i+1) % n, (i+1) % n+n, i+n])
        polygonalFaces[0][i] = i
        polygonalFaces[1][i] = n+i
    faces.append(polygonalFaces[0])
    faces.append(polygonalFaces[1])

    return verts, faces


def main():
    cells = {}
    i = 5
    filename = str("prism_"+str(i).zfill(3)+".vtk")
    points, faces = prism(i)
    triangles = []
    triangles.append([0, 1, 5])
    triangles.append([5, 6, 1])
    cells["triangle"] = np.array(triangles)
    cells["quad"] = np.array(faces[1:i])
    cells["polygon"] = np.array(faces[i:i+2])
    meshio.write_points_cells(
        filename,
        points,
        cells,
        file_format="vtk-ascii"
        )
    print(cells)
    cells.clear()

    points, elements = readMesh(filename)
    print(elements)
```

File: meshio/vtk_io.py (hunk 4 above)
Comment (nschloe) [reply meshio/vtk_io.py]: Ah, same misunderstanding as above. When I read `in` I think lists. Please also replace by `key[:7] == "polygon"`.

**Comment by nschloe** [comment]
Can you allow edits from maintainers? https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by eolianoe** [comment]
> Can you allow edits from maintainers? https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

Done

**Comment by nschloe** [comment]
Okay so, the tests are failing if different polygons are present in a mesh. Please fix! :smiley:

**Comment by eolianoe** [comment]
I add polygons to the legacy reader and tests are running fine, do not hesitate to tell me about bad coding style

**Comment by nschloe** [comment]
Merged & released. Thanks for a nice contribution!

**Comment by eolianoe** [comment]
You're welcome, thanks for the feedback and the great library !

File: meshio/helpers.py (hunk 1 above)
Comment (MuellerSeb) [reply meshio/helpers.py]: Why not use the pythonic way: ``key.startswith("polygon")``?

Comment (eolianoe) [reply meshio/helpers.py]: Because my python is quite new, what's your opinion @nschloe?

**Comment by nschloe** [comment]
Don't mind either way.

**Pull request merged**